### ISSUE TITLE
History has different length on two machines after `atuin sync`

### ISSUE THREAD

**Issue by mrnugget** [atuinsh/atuin#1596]
I have two machines logged-in to the same atuin account. Both used `atuin import` to import their respective `~/.zsh_history` into atuin.

What I expected: both machines would end up with the same history shared history, synced across machines.

What I see: one machine has a longer history than the other, as if not everything from its history was synced.

### Machine A

```
φ hostname
m1max
φ atuin history list --format "{host} - {time} - {command}" | head -n 10
greyunixbox - 2019-08-23 08:39:21 - rg 'Commit\(' cmd/frontend/graphqlbackend
greyunixbox - 2019-08-23 09:01:30 - gd > ~/work/git_author_optional.diff
greyunixbox - 2019-08-23 09:17:49 - git co -b core/all-sync-errors
greyunixbox - 2019-08-23 09:30:35 - go test -run 'TestSyncer_Sync/sourcer_error_aborts_syn' ./cmd/repo-updater/...
greyunixbox - 2019-08-23 09:36:59 - go test -run 'TestSyncer_Sync' ./cmd/repo-updater/...
greyunixbox - 2019-08-23 10:53:54 - rg 'CloningStatusMessage'
greyunixbox - 2019-08-23 10:57:08 - ruplacer SyncErrorStatusMessage ExternalServiceSyncError
greyunixbox - 2019-08-23 10:57:13 - ruplacer SyncErrorStatusMessage ExternalServiceSyncError --go
greyunixbox - 2019-08-23 10:57:25 - rg SyncError
greyunixbox - 2019-08-23 10:57:43 - rg '[^Last]SyncError'
φ atuin history list | wc -l
   27164
```

### Machine B

```
~ φ hostname
greyunixbox
~ φ atuin history list --format "{host} - {time} - {command}" | head -n 10
greyunixbox - 2019-04-25 16:32:11 - git clone https://github.com/sourcegraph/sourcegraph
greyunixbox - 2019-04-25 17:04:00 - npm install -g yarn
greyunixbox - 2019-04-25 17:04:59 - which tmux
greyunixbox - 2019-04-25 17:09:01 - mv .envrc .env
greyunixbox - 2019-04-25 20:29:46 - curl localhost:3081
greyunixbox - 2019-04-25 20:35:11 - brew uninstall node
greyunixbox - 2019-04-25 20:35:15 - rm -rf .node-gyp
greyunixbox - 2019-04-25 20:35:18 - rm -rf .npm
greyunixbox - 2019-04-25 20:35:48 - asdf plugin-add nodejs
greyunixbox - 2019-04-25 20:36:02 - asdf golang list
~ φ atuin history list | wc -l
29308
```

---

So it looks like `m1max` _has_ imported the history from `greyunixbox`, but it's missing 4 months at the start of the history?

Is that because syncing might still be ongoing and is eventually consistent? Or is there another reason? Can I do something to debug this?

**Comment by ellie** [comment]
Hey!

Just a few quick things to check

1. Are you self hosting the Atuin server? 
2. What is the output of `atuin status`, on both machines? It should show local/remote history counts
3. If you run `atuin sync -f` on both machines, does it sort itself out?

> Is that because syncing might still be ongoing and is eventually consistent?

It is eventually consistent, though shouldn't still be ongoing - unless you're uploading a lot, initial syncs usually complete in <1min.

> Can I do something to debug this?

`ATUIN_LOG=debug atuin sync` can be super helpful, although very very chatty

**Comment by mrnugget** [comment]
`atuin status` on `greyunixbox`:

```
~ φ hostname
greyunixbox
~ φ atuin status
Atuin v17.2.1 - Build rev 9f79a34a9dea207df7184972fa29f4fe5ec41b27

[Local]
Sync frequency: 10m
Last sync: 2024-01-21 9:31:13.844998715 +00:00:00
History count: 27423
Deleted history count: 0

[Remote]
Address: https://api.atuin.sh
Username: mrnugget
History count: 27829
```

`atuin status` on `m1max`:

```
~ φ hostname
m1max
~ φ atuin status
Atuin v17.2.1 - Build rev

[Local]
Sync frequency: 10m
Last sync: 2024-01-21 9:31:04.521809 +00:00:00
History count: 25681
Deleted history count: 0

[Remote]
Address: https://api.atuin.sh
Username: mrnugget
History count: 27829
```

After `atuin sync -f` on `greyunixbox` (the one with more history) I get this:

```
~ φ atuin sync -f
0/0 up/down to record store
Sync complete! 28132 items in history database, force: true
~ φ atuin status
Atuin v17.2.1 - Build rev 9f79a34a9dea207df7184972fa29f4fe5ec41b27

[Local]
Sync frequency: 10m
Last sync: 2024-01-21 9:33:00.43116819 +00:00:00
History count: 28133
Deleted history count: 0

[Remote]
Address: https://api.atuin.sh
Username: mrnugget
History count: 27829
```

So it seems like there's stuff it can't import?

**Comment by ellie** [comment]
> So it seems like there's stuff it can't import?

Sure does - sometimes this happens when there's something very large in your history (ie, when encrypted + with metadata it's > a few mb). I don't think that's happening here though, and there's nothing in my logs + metrics to show that that's the case.

Does

```
ATUIN_LOG=atuin_client::sync=debug atuin sync -f
```

On either/both machines show anything enlightening? I've double checked server logs and there's nothing suspicious there.

There's always a chance when importing zsh_history files that _something_ has been missed/gone wrong 🤔 Is new data syncing ok?

**Comment by mrnugget** [comment]
Hmmm, it just seems to stop before uploading the earliest 3 months:

```
~ φ ATUIN_LOG=atuin_client::sync=debug atuin sync -f
0/0 up/down to record store
[2024-01-21T13:48:08Z DEBUG atuin_client::sync] starting sync upload
[2024-01-21T13:48:08Z DEBUG atuin_client::sync] remote has 28623, we have 28876
[2024-01-21T13:48:09Z DEBUG atuin_client::sync] upload cursor: 2023-10-08 6:19:17.466 +00:00:00
[2024-01-21T13:48:10Z DEBUG atuin_client::sync] upload cursor: 2023-08-02 4:33:32.981 +00:00:00
[2024-01-21T13:48:10Z DEBUG atuin_client::sync] upload cursor: 2023-07-11 4:01:58.1 +00:00:00
[2024-01-21T13:48:11Z DEBUG atuin_client::sync] upload cursor: 2022-12-27 9:46:36.066 +00:00:00
[2024-01-21T13:48:12Z DEBUG atuin_client::sync] upload cursor: 2022-08-04 4:52:31.968 +00:00:00
[2024-01-21T13:48:13Z DEBUG atuin_client::sync] upload cursor: 2022-01-24 10:33:40.923 +00:00:00
[2024-01-21T13:48:13Z DEBUG atuin_client::sync] upload cursor: 2021-10-22 16:00:03.823 +00:00:00
[2024-01-21T13:48:14Z DEBUG atuin_client::sync] upload cursor: 2021-09-15 15:08:41.725 +00:00:00
[2024-01-21T13:48:15Z DEBUG atuin_client::sync] upload cursor: 2021-07-23 9:21:11.626 +00:00:00
[2024-01-21T13:48:16Z DEBUG atuin_client::sync] upload cursor: 2021-06-16 5:52:23.227 +00:00:00
[2024-01-21T13:48:16Z DEBUG atuin_client::sync] upload cursor: 2021-03-26 10:49:13.777 +00:00:00
[2024-01-21T13:48:17Z DEBUG atuin_client::sync] upload cursor: 2020-12-01 14:05:58.677 +00:00:00
[2024-01-21T13:48:18Z DEBUG atuin_client::sync] upload cursor: 2020-10-26 12:50:21.577 +00:00:00
[2024-01-21T13:48:19Z DEBUG atuin_client::sync] upload cursor: 2020-09-20 6:19:18.477 +00:00:00
[2024-01-21T13:48:20Z DEBUG atuin_client::sync] upload cursor: 2020-08-19 9:58:20.377 +00:00:00
[2024-01-21T13:48:20Z DEBUG atuin_client::sync] upload cursor: 2020-06-10 12:00:36.277 +00:00:00
[2024-01-21T13:48:21Z DEBUG atuin_client::sync] upload cursor: 2020-04-25 14:22:39.177 +00:00:00
[2024-01-21T13:48:22Z DEBUG atuin_client::sync] upload cursor: 2020-03-18 13:42:52.077 +00:00:00
[2024-01-21T13:48:23Z DEBUG atuin_client::sync] upload cursor: 2020-02-14 13:41:30.977 +00:00:00
[2024-01-21T13:48:23Z DEBUG atuin_client::sync] upload cursor: 2020-01-14 11:23:42.877 +00:00:00
[2024-01-21T13:48:24Z DEBUG atuin_client::sync] upload cursor: 2019-11-22 9:55:13.777 +00:00:00
[2024-01-21T13:48:25Z DEBUG atuin_client::sync] upload cursor: 2019-10-23 13:18:00.677 +00:00:00
[2024-01-21T13:48:26Z DEBUG atuin_client::sync] upload cursor: 2019-09-06 8:01:51.577 +00:00:00
[2024-01-21T13:48:27Z DEBUG atuin_client::sync] upload cursor: 2019-07-19 13:48:18.477 +00:00:00
[2024-01-21T13:48:27Z DEBUG atuin_client::sync] starting sync download
[2024-01-21T13:48:29Z DEBUG atuin_client::sync] sync downloaded 540
Sync complete! 29416 items in history database, force: true
```

But when I look at the `.zsh_history` I don't see any standing out there: no format change or anything like that.

Then I ran it two more times and it uploaded more data:

```
~ φ ATUIN_LOG=atuin_client::sync=debug atuin sync -f
0/0 up/down to record store
[2024-01-21T13:53:16Z DEBUG atuin_client::sync] starting sync upload
[2024-01-21T13:53:16Z DEBUG atuin_client::sync] remote has 29957, we have 30714
[2024-01-21T13:53:17Z DEBUG atuin_client::sync] upload cursor: 2023-12-13 13:32:04.99 +00:00:00
[2024-01-21T13:53:18Z DEBUG atuin_client::sync] upload cursor: 2023-11-03 13:19:37.115 +00:00:00
[2024-01-21T13:53:18Z DEBUG atuin_client::sync] upload cursor: 2023-09-08 17:30:25.255 +00:00:00
[2024-01-21T13:53:19Z DEBUG atuin_client::sync] upload cursor: 2023-07-14 18:18:26.838 +00:00:00
[2024-01-21T13:53:20Z DEBUG atuin_client::sync] upload cursor: 2023-04-25 4:23:47.626 +00:00:00
[2024-01-21T13:53:21Z DEBUG atuin_client::sync] upload cursor: 2022-11-13 12:07:54.704 +00:00:00
[2024-01-21T13:53:21Z DEBUG atuin_client::sync] upload cursor: 2022-06-15 4:35:40.635 +00:00:00
[2024-01-21T13:53:22Z DEBUG atuin_client::sync] upload cursor: 2021-12-22 10:44:43.561 +00:00:00
[2024-01-21T13:53:23Z DEBUG atuin_client::sync] upload cursor: 2021-10-08 8:27:02.463 +00:00:00
[2024-01-21T13:53:24Z DEBUG atuin_client::sync] upload cursor: 2021-09-08 9:56:56.363 +00:00:00
[2024-01-21T13:53:24Z DEBUG atuin_client::sync] upload cursor: 2021-07-13 12:08:45.269 +00:00:00
[2024-01-21T13:53:25Z DEBUG atuin_client::sync] upload cursor: 2021-06-01 4:52:32.515 +00:00:00
[2024-01-21T13:53:26Z DEBUG atuin_client::sync] upload cursor: 2021-03-02 6:13:37.415 +00:00:00
[2024-01-21T13:53:27Z DEBUG atuin_client::sync] upload cursor: 2020-11-19 8:35:16.315 +00:00:00
[2024-01-21T13:53:27Z DEBUG atuin_client::sync] upload cursor: 2020-10-15 7:29:54.215 +00:00:00
[2024-01-21T13:53:28Z DEBUG atuin_client::sync] upload cursor: 2020-09-11 13:24:14.115 +00:00:00
[2024-01-21T13:53:29Z DEBUG atuin_client::sync] upload cursor: 2020-07-22 14:59:18.015 +00:00:00
[2024-01-21T13:53:30Z DEBUG atuin_client::sync] upload cursor: 2020-05-22 5:09:57.915 +00:00:00
[2024-01-21T13:53:30Z DEBUG atuin_client::sync] upload cursor: 2020-04-13 10:57:29.815 +00:00:00
[2024-01-21T13:53:31Z DEBUG atuin_client::sync] upload cursor: 2020-03-02 12:55:28.715 +00:00:00
[2024-01-21T13:53:32Z DEBUG atuin_client::sync] upload cursor: 2020-01-31 14:58:39.615 +00:00:00
[2024-01-21T13:53:33Z DEBUG atuin_client::sync] upload cursor: 2019-12-13 14:42:48.515 +00:00:00
[2024-01-21T13:53:34Z DEBUG atuin_client::sync] upload cursor: 2019-11-15 12:29:09.415 +00:00:00
[2024-01-21T13:53:34Z DEBUG atuin_client::sync] upload cursor: 2019-10-04 12:10:02.315 +00:00:00
[2024-01-21T13:53:35Z DEBUG atuin_client::sync] upload cursor: 2019-08-23 13:30:55.215 +00:00:00
[2024-01-21T13:53:36Z DEBUG atuin_client::sync] upload cursor: 2019-07-11 8:59:06.115 +00:00:00
[2024-01-21T13:53:37Z DEBUG atuin_client::sync] upload cursor: 2019-05-31 15:25:25.015 +00:00:00
[2024-01-21T13:53:37Z DEBUG atuin_client::sync] starting sync download
[2024-01-21T13:53:40Z DEBUG atuin_client::sync] sync downloaded 1121
Sync complete! 31835 items in history database, force: true
```

Here it went to 2019-05-31. Then, next sync:

```
~ φ ATUIN_LOG=atuin_client::sync=debug atuin sync -f
0/0 up/down to record store
[2024-01-21T13:54:32Z DEBUG atuin_client::sync] starting sync upload
[2024-01-21T13:54:32Z DEBUG atuin_client::sync] remote has 30870, we have 31838
[2024-01-21T13:54:33Z DEBUG atuin_client::sync] upload cursor: 2024-01-09 14:34:57.711 +00:00:00
[2024-01-21T13:54:34Z DEBUG atuin_client::sync] upload cursor: 2023-12-14 5:21:10.012 +00:00:00
[2024-01-21T13:54:35Z DEBUG atuin_client::sync] upload cursor: 2023-11-03 13:41:36.139 +00:00:00
[2024-01-21T13:54:35Z DEBUG atuin_client::sync] upload cursor: 2023-09-09 4:27:23.279 +00:00:00
[2024-01-21T13:54:36Z DEBUG atuin_client::sync] upload cursor: 2023-07-17 4:39:22.9 +00:00:00
[2024-01-21T13:54:37Z DEBUG atuin_client::sync] upload cursor: 2023-05-14 7:15:28.647 +00:00:00
[2024-01-21T13:54:38Z DEBUG atuin_client::sync] upload cursor: 2022-11-17 5:24:08.728 +00:00:00
[2024-01-21T13:54:39Z DEBUG atuin_client::sync] upload cursor: 2022-06-17 5:15:21.659 +00:00:00
[2024-01-21T13:54:39Z DEBUG atuin_client::sync] upload cursor: 2021-12-23 7:51:59.585 +00:00:00
[2024-01-21T13:54:40Z DEBUG atuin_client::sync] upload cursor: 2021-10-11 4:51:58.485 +00:00:00
[2024-01-21T13:54:41Z DEBUG atuin_client::sync] upload cursor: 2021-09-08 14:32:25.387 +00:00:00
[2024-01-21T13:54:42Z DEBUG atuin_client::sync] upload cursor: 2021-07-14 8:19:59.293 +00:00:00
[2024-01-21T13:54:42Z DEBUG atuin_client::sync] upload cursor: 2021-06-02 5:23:24.538 +00:00:00
[2024-01-21T13:54:43Z DEBUG atuin_client::sync] upload cursor: 2021-03-02 13:14:34.439 +00:00:00
[2024-01-21T13:54:44Z DEBUG atuin_client::sync] upload cursor: 2020-11-19 14:20:45.339 +00:00:00
[2024-01-21T13:54:45Z DEBUG atuin_client::sync] upload cursor: 2020-10-15 7:55:45.239 +00:00:00
[2024-01-21T13:54:45Z DEBUG atuin_client::sync] upload cursor: 2020-09-11 15:01:56.139 +00:00:00
[2024-01-21T13:54:46Z DEBUG atuin_client::sync] upload cursor: 2020-07-23 14:51:19.039 +00:00:00
[2024-01-21T13:54:47Z DEBUG atuin_client::sync] upload cursor: 2020-05-25 9:03:23.939 +00:00:00
[2024-01-21T13:54:48Z DEBUG atuin_client::sync] upload cursor: 2020-04-13 11:27:02.839 +00:00:00
[2024-01-21T13:54:48Z DEBUG atuin_client::sync] upload cursor: 2020-03-03 14:56:11.739 +00:00:00
[2024-01-21T13:54:49Z DEBUG atuin_client::sync] upload cursor: 2020-01-31 16:15:54.639 +00:00:00
[2024-01-21T13:54:50Z DEBUG atuin_client::sync] upload cursor: 2019-12-17 8:24:48.539 +00:00:00
[2024-01-21T13:54:51Z DEBUG atuin_client::sync] upload cursor: 2019-11-15 14:14:27.439 +00:00:00
[2024-01-21T13:54:51Z DEBUG atuin_client::sync] upload cursor: 2019-10-07 8:43:50.339 +00:00:00
[2024-01-21T13:54:52Z DEBUG atuin_client::sync] upload cursor: 2019-08-26 9:22:22.239 +00:00:00
[2024-01-21T13:54:53Z DEBUG atuin_client::sync] upload cursor: 2019-07-12 7:01:21.139 +00:00:00
[2024-01-21T13:54:54Z DEBUG atuin_client::sync] upload cursor: 2019-06-01 9:33:57.039 +00:00:00
[2024-01-21T13:54:55Z DEBUG atuin_client::sync] upload cursor: 2019-04-25 16:32:11.001 +00:00:00
[2024-01-21T13:54:55Z DEBUG atuin_client::sync] starting sync download
[2024-01-21T13:54:58Z DEBUG atuin_client::sync] sync downloaded 49
Sync complete! 31887 items in history database, force: true
```

Then I synced the other machines too and now I think they all line up, because they all show the same result when doing `atuin history list | head -n 10`.

So I think we're good now, but I had to manually sync multiple times. Maybe an indexing or off-by-one error? (I'm saying this because if _I_ wrote code for syncing a timeline you can bet there would be off-by-one bugs in it :))

**Comment by ellie** [comment]
Glad it's all sorted! Sorry for the issues there :/ 

> Maybe an indexing or off-by-one error?

Hmmm, maybe. The issue we tend to have with imports is that history sync relies on time (why did I do this 🙃), and while Atuin tends to record history with the correct timestamps, sometimes either the importer or shell-recorded timestamp can be off. You can see it paging through timestamps in the logs there.

There's a new sync that's been WIP for a long while that's now on `main`. It should be far more robust, but isn't yet ready for general usage. After a bunch more testing it'll be out as opt-in! (and then eventually the default). It doesn't depend on timestamps or anything like that 🙏

**Comment by keichan34** [comment]
I had a similar problem. In my case, I had one machine that was on the original sync protocol, and another that was newly set up, that defaulted to the new sync v2 protocol. They had completely different sync histories, no matter how many times I ran sync -f with the debug logging detailed in this issue. 

I solved the problem by [enabling sync v2](https://docs.atuin.sh/configuration/config/#sync), and re-running `atuin sync -f` on both machines.
